### PR TITLE
docs: add installation guide

### LINKED ISSUE #810: [📑 Docs]: Provide new doc about installation and usage
### What Dev Docs changes are you proposing?

We miss the following docs in current outline: Installation and Usage

What do you think, we need one, two or 3 separate docs?

Nevertheless technical input 👇🏼 

## Installation

I think it makes sense to also add new document called `Installation` where we explain how to install the generator (and also update) and what are technical requirements. Better separate than mixing it with the Intro doc. Even though we know `ag` CLI is going away soon, under `asyncapi` CLI the way it works will not be much different, so simple update to docs will need to be done, not much work

technical input -> https://github.com/asyncapi/generator#using-from-the-command-line-interface-cli

also I think https://github.com/asyncapi/generator#generator-version-vs-template-version should be here. I'm not sure if it should be separate document.

## Using Generator in CLI

We need to explain how to use generator in the CLI

Actually all that we have in the readme must be here in docs. Because ultimate goal is to have all generator docs in one location, and in readme long term we will just link to docs.

Don't worry, we are migrating to new CLI but commands will not change much. Later simple update with search and replace will be possible.

## Using Generator as a library

Readme already has nice simple example on how o use generator npm module. Just the thing is that it is linking to https://github.com/asyncapi/generator/blob/master/docs/api.md

The `api.md` is generated using `npm run docs` and ultimate goal -> we also have it rendered in asyncapi.com. So please, with this issue, also add `api.md` to this readme -> https://github.com/asyncapi/generator/blob/master/docs/README.md

cc @pratik2315 @Florence-Njeri

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by Florence-Njeri**
**Description**
This is the ag CLI installation and upgrade guide.

**Related issue(s)**
`Fixes #810`

### CODE REVIEW

File: docs/installation-guide.md
Comment (derberg) [inline docs/installation-guide.md:27]: ups, we did not update readme, we actually also test on node 16 and newer npm, so we are sure that newer version work like a charm, we can remove this note

**Review by derberg — CHANGES_REQUESTED** [review]
@Florence-Njeri 

We are missing some content here.

So the thing is that there are 2 ways (actually 3 😄 ) to install AsyncAPI generator

- generator is a library that anyone can also use in their Nodejs applications. Like we do in our server-api -> https://github.com/asyncapi/server-api/blob/master/package.json#L49 
So you use it as library, not its CLI. You install it like `npm install @asyncapi/generator`, not with `-g` flag, as you do not install it globally but only inside your project. Sample code -> https://github.com/asyncapi/server-api/blob/master/src/services/generator.service.ts#L18-L29. **We need this covered in this document too, I mean info that it is a library and how to install in a project (same npm and nodejs restrictions)
- generator also contains cli. Atm PR to introduce generator in `@asyncapi/cli` is still not merged. So I suggest, for now, forget about it. Do not mention how to install `@asyncapi/cli`  (also remove MacOS instructions as they are valid only for AsyncAPI CLI, not generator CLI). So focus only on `@asyncapi/generator` please

> (actually 3 😄 )

it would be nifty to mention that if someone wants to use Generator CLI in CI/CD to automate generation (many people do it), then whey should just install generator CLI in their CI/CD, but in case of GitHub Actions, they can use https://github.com/marketplace/actions/generator-for-asyncapi-documents


---
I hope I'm not confusing, please lemme know if you need more help

File: docs/installation-guide.md
Comment (Florence-Njeri) [inline docs/installation-guide.md:15]: @derberg I wanted to link the 2 ways to install the Generator here but VS code kept auto-generating the table of contents instead. I just wanted it to be:

1.  [Using the AsyncAPI Generator CLI](#1-using-the-asyncapi-generator-cli) and 
2.  [Use the Generator Library In Your Node.js Apps](#2-use-the-generator-library-in-your-nodejs-apps)
Not sure how to achieve this.

File: docs/installation-guide.md
Comment (Florence-Njeri) [inline docs/installation-guide.md:74]: I have some doubts about how I mention the prerequisites here. I'm not sure if it should be above the small sentence I have or below.

File: docs/versioning.md
Comment (Florence-Njeri) [inline docs/versioning.md:4]: This doc is about **the generator version vs the template version** but I found most of what I explained is about the generator version, not the template version. This is one thing I think can be improved in this doc. Maybe you can give me some pointers.

**Comment by Florence-Njeri** [comment]
Hey @alequetzalli, @derberg, @pratik2315, @Annysah, @starlightknown, @nelsonmic, @thulieblack the installation guide is ready for review. I'll appreciate any feedback/criticism I get 😄

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:6]:
```suggestion
There are various ways you can use the Generator library to generate whatever you want in your event-driven architecture apps. Find your preferred method below:
```

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:27]:
```suggestion
## Generator CLI
```

When writing **headers** in docs, focus on being **succinct**. Avoid unnecessary terms, verbs,  or articles _(i.e. the, of, etc)_. 

Example: `asyncapi.com/docs/tools/generator#generator-cli`.

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:56]:
```suggestion
### Uninstall
```

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:64]:
```suggestion
## Generator Library in Node.js Apps
```

This is about how the URL and your page outline make sense and seem cleaner this way:
`asyncapi.com/docs/tools/generator#generator-library-in-nodejs-apps`

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:1]: @Florence-Njeri Hey, so a reminder again that we use sentence capitalization in headers for Docs. (I keep having to update your headers girl 😜😜😜😜)

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:30]: great header, for both TW & SEO!

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:10]: great header, for both TW & SEO!

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:7]:
```suggestion
- [Generator CLI](#generator-cli)
```

Not sure if this URL works now, but I wanted to make sure we shorten this to be succinct

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:25]:
```suggestion
If you have the correct versions installed, proceed to the CLI installation guide below. Otherwise, upgrading the npm or node version is lower than the recommended versions specified above.
```

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:23]:
```suggestion
If you don't have either Node or Npm installed, use the [official node.js installer](https://nodejs.org/en/download/).
```

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:11]:
```suggestion
Before you install and use the AsyncAPI CLI and the generator library, ensure you meet the prerequisites below, then [install the CLI](#installation).
```

File: docs/installation-guide.md
Comment (Florence-Njeri) [reply docs/installation-guide.md]: Gotcha! This makes so much sense, when writing I wasn't putting into consideration how the links to the different sections will look like or even SEO! Thank you for enlightening me

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:28]:
```suggestion
The CLI tool allows you to generate message-based API boilerplate code, documentation, or anything else you need as long as you specify it in your [template](template.md). To use the Generator via the CLI, you will need to install the AsyncApi Generator CLI tool in your local development environment.
```

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:32]:
```suggestion
The recommended method to interact with the generator tool is its CLI. To use the CLI, install it  using the following `npm` command:
```

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:42]:
```suggestion
> Each [community developed template](https://github.com/search?q=topic%3Aasyncapi+topic%3Agenerator+topic%3Atemplate) is dependent on a certain version of the generator for it to work correctly. Before you install the generator CLI, check the templates' `package.json` for the version of the Generator CLI your template is compatible with. To learn more about the generator version and how it correlates to the template you use check the [versioning docs](versioning.md).
```

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:46]:
```suggestion
1. You have the generator tool installed but want to use the latest released features. To upgrade to the latest version, use the command below:
```

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:50]:
```suggestion
2. If your template isn't compatible with the latest generator version, you can update it to a specific version of the generator. Check the [version you need](https://github.com/asyncapi/generator/releases) and specify the version you want by using the **@** symbol as shown in the command below:
```

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:62]:
```suggestion
> :memo: **Note:**  If you want to use the Generator CLI in your CI/CD pipeline to automate whatever you generate for your event-driven architecture apps, install the AsyncAPI Generator CLI in your pipeline. If you are using GitHub Actions, use [Github Actions for Generator](https://github.com/marketplace/actions/generator-for-asyncapi-documents).
```

File: docs/installation-guide.md
Comment (alequetzalli) [inline docs/installation-guide.md:65]:
```suggestion
Use the generator library in your Node.js projects by installing it via the following command: `npm install @asyncapi/generator`.
```

File: docs/versioning.md
Comment (alequetzalli) [inline docs/versioning.md:6]:
```suggestion
The generator tool generates whatever you want, as long as it can be defined in a template, based on the [AsyncAPI file](asyncapi-file.md). On the other hand, a **template** is a file or group of files that specify the kind of output you expect from using the generator's features. For example, you may want to use the [nodejs template](https://github.com/asyncapi/nodejs-template) to generate boilerplate code for your message-based APIs.
```

File: docs/versioning.md
Comment (alequetzalli) [inline docs/versioning.md:8]:
```suggestion
Templates are dependent on the generators' features. For example, the template you want to use may be compatible with the latest generator version but incompatible with the previous versions. Check the configuration file or ReadME of the template to see the version of the generator it supports. The generator has an `isTemplateCompatible` function that checks if the template is compatible with the version of the generator you want to use. If the template isn't compatible, you will see a terminal error output similar to the following:
```

File: docs/versioning.md
Comment (alequetzalli) [inline docs/versioning.md:16]:
```suggestion
It is better to lock a specific version of the template and the Generator if you plan to use the Generator CLI and a particular template in production. The differences between using the version of the Generator CLI you have installed and locking a certain version on production are demonstrated in the following code snippets.
```

File: docs/versioning.md
Comment (alequetzalli) [inline docs/versioning.md:18]:
```suggestion
Generate HTML with the latest AsyncAPI Generator CLI using the html-template.
```

File: docs/versioning.md
Comment (alequetzalli) [inline docs/versioning.md:24]:
```suggestion
Generate HTML using a particular version of the AsyncAPI Generator CLI using the html-template.
```

**Review by alequetzalli — CHANGES_REQUESTED** [review]
Here's your next review round, @Florence-Njeri ! 😄

File: docs/installation-guide.md
Comment (Florence-Njeri) [reply docs/installation-guide.md:1]: I'm sorry, I promise it won't happen again 😅

File: docs/versioning.md
Comment (derberg) [inline docs/versioning.md:6]:
```suggestion
The generator tool generates whatever you want, as long as it can be defined in a template based on the [AsyncAPI file](asyncapi-file.md). On the other hand, a **template** is a file or group of files that specify the kind of output you expect from using the generator's features. For example, you may want to use the [NodeJS template](https://github.com/asyncapi/nodejs-template) to generate boilerplate code for your message-based APIs.
```

`defined in a template based` - maybe link to template too?

File: docs/versioning.md
Comment (derberg) [inline docs/versioning.md:14]:
```suggestion
> Use the following command to check the version of the AsyncAPI CLI you have installed;  `asyncapi --version`
```

File: docs/versioning.md
Comment (derberg) [inline docs/versioning.md:16]:
```suggestion
It is better to lock a specific version of the template and the Generator if you plan to use the AsyncAPI CLI and a particular template in production. The differences between using the version of the AsyncAPI CLI you have installed and locking a certain version on production are demonstrated in the following code snippets.
```

File: docs/versioning.md
Comment (derberg) [inline docs/versioning.md:21]:
```suggestion
asyncapi generate fromTemplate asyncapi.yaml @asyncapi/html-template -o ./docs
```

File: docs/versioning.md
Comment (derberg) [inline docs/versioning.md:27]:
```suggestion
npm install -g @asyncapi/cli@0.20.0
```

File: docs/versioning.md
Comment (derberg) [inline docs/versioning.md:28]:
```suggestion
asyncapi generate fromTemplate asyncapi.yaml @asyncapi/html-template@0.7.0 -o ./docs
```

File: docs/versioning.md
Comment (derberg) [inline docs/versioning.md:18]:
```suggestion
Generate HTML with the latest AsyncAPI CLI using the html-template.
```

**Review by derberg — APPROVED** [review]
Sorry for so many fixes, but yeah `asyncapi generate fromTemplate` was released.

I made all comments as suggestions, so you can easily apply them from the UI and push all, and then I will have a final review just to make sure all is good

**Review by derberg — CHANGES_REQUESTED** [review]
sorry, I should not approve 😅

File: docs/installation-guide.md
Comment (Florence-Njeri) [reply docs/installation-guide.md:7]: Thank you @derberg

File: docs/versioning.md
Comment (alequetzalli) [inline docs/versioning.md:16]:
```suggestion
It is better to lock a specific version of the template and the generator if you plan to use the AsyncAPI CLI and a particular template in production. The differences between using the version of the AsyncAPI CLI you have installed and locking a certain version on production are demonstrated in the following code snippets.
```

Consistency is KEY 🔑 . You haven't been capitalizing `generator` mentions so far, so let's make them all match.

**Review by alequetzalli — COMMENTED** [review]
Just one tiny commit suggestion fix @Florence-Njeri  , but I think this looks good to go! 

@derberg do you approve of this too? 😬✌🏽😄

File: docs/installation-guide.md
Comment (derberg) [inline docs/installation-guide.md:52]:
```suggestion
* If your template isn't compatible with the latest generator version, you can update it to a specific version of the generator. Check the [version you need](https://github.com/asyncapi/cli/releases) and specify the version you want by using the **@** symbol as shown in the command below:
```

**Review by derberg — CHANGES_REQUESTED** [review]
@alequetzalli all good, spotted just one tiny error. 

@Florence-Njeri now just remove relevant content sections from https://github.com/asyncapi/generator/blob/master/README.md and link to your new documents. So in case of `Installation` section in the readme, it should stay there, but just put there a short paragraph-link to separate document

File: docs/versioning.md
Comment (alequetzalli) [reply docs/versioning.md:16]: @Florence-Njeri wait, is this commit added? it looks like it's pending 🤔

File: docs/installation-guide.md
Comment (alequetzalli) [reply docs/installation-guide.md:45]: Interesting ! I think this is very long and verbose, without telling us much at all. However, I DO think we should include this in this installation section because documenting diff OS installation is a best practice. Here is a standard example from Gastby that shows how this can look like... 

Screen Shot 2022-10-19 at 2 07 36 PM

## My feedback suggestion
Florence, I'd prob shorten this and improve it by making a tiny section per top two OS minimum:

```
#### MacOS 
You can install in MacOS by using brew: `brew install asyncapi`. 

#### Linux 
You can install in Linux by using `dpkg`, a package manager for debian:
1. 'curl -OL https://github.com/asyncapi/cli/releases/latest/download/asyncapi.deb`
2. `sudo dpkg -i asyncapi.deb`

#### Other Operating Systems
For further installation instructions for different operating systems, read the [`AsyncAPI CLI` readME file](https://github.com/asyncapi/cli#installation)
```

Comment (derberg) [reply docs/installation-guide.md:45]: ok, and overtime we can always ask volunteers to update doc as first time contribution, of there are more methods available 👍🏼

**Review by derberg — CHANGES_REQUESTED** [review]
few minors

also please add these new docs here https://github.com/asyncapi/generator/blob/master/docs/README.md

After Intro, and 2nd and 3rd doc

**Review by derberg — APPROVED** [review]
LGTM 🚀 
@alequetzalli wanna have a final look before I merge?

File: docs/installation-guide.md
Comment (derberg) [inline docs/installation-guide.md:53]:
```suggestion
For further installation instructions for different operating systems, read the [AsyncAPI CLI documentation](https://github.com/asyncapi/cli#installation).
```

**Review by derberg — DISMISSED** [review]
lgtm 🚀 

@alequetzalli wanna approve before merge?

**Review by alequetzalli — DISMISSED** [review]
MERGING TIME BABY 🚢 🚢 🚢 🚢 🚢 🚢

**Review by derberg — APPROVED** [review]

**Pull request merged**
